### PR TITLE
DFE interval with bed file

### PR DESCRIPTION
**Pull request by izabelcavassim**
Adding --dfe-bed-file as discussed in #1112

### CODE REVIEW

**Comment by izabelcavassim** [comment]
@andrewkern

File: stdpopsim/cli.py [hunk 1]
```diff
@@ -722,6 +735,11 @@ def run_simulation(args):
                 annot = species.get_annotations(args.dfe_annotation)
                 intervals = annot.get_chromosome_annotations(args.chromosome)
                 intervals_summary_str = f"{annot.id} elements on {args.chromosome}"
+            elif args.dfe_bed_file is not None:
```
Comment (andrewkern) [inline stdpopsim/cli.py:738]: we're going to want to do some testing to make sure this option isn't called with `--dfe-interval` or `--dfe-annotation`

Comment (izabelcavassim) [reply stdpopsim/cli.py:738]: Oh I just saw this now, yes I can do that!

**Comment by izabelcavassim** [comment]
Can you have a look at it @petrelharp

File: stdpopsim/cli.py (hunk 1 above)
Comment (izabelcavassim) [reply stdpopsim/cli.py:753]: added

File: stdpopsim/cli.py [hunk 2]
```diff
@@ -608,7 +608,20 @@ def add_simulate_species_parser(parser, species):
         metavar="",
         help=interval_help,
     )
-
+    bed_help = (
+        "A bed file specifing the interval where selection (given a DFE) is simulated. "
```
Comment (petrelharp) [inline stdpopsim/cli.py:612]:
```suggestion
        "A bed file specifing the intervals where selection (given a DFE) is simulated. "
```

File: stdpopsim/cli.py [hunk 3]
```diff
@@ -608,7 +608,20 @@ def add_simulate_species_parser(parser, species):
...
+        "Non-overlaping intervals belong to the same chromosome are required. "
+        "If no interval is specified, "
+        "selection is simulated across the entire region (contig). "
+        "Check also --dfe-interval and --dfe-annotation."  # or something on those lines
```
Comment (petrelharp) [inline stdpopsim/cli.py:616]:
```suggestion
        "Non-overlaping intervals belonging to the same chromosome are required. "
        "If no interval is specified, "
        "selection is simulated across the entire contig. "
        "See also --dfe-interval and --dfe-annotation."
```

File: stdpopsim/cli.py [hunk 4]
```diff
@@ -708,20 +721,40 @@ def run_simulation(args):
                     "A DFE annotation has been assigned without a DFE. "
                     "Please specify a DFE."
                 )
+            elif args.dfe_bed_file is not None:
```
Comment (petrelharp) [inline stdpopsim/cli.py:724]: This and the one above should probably be `if`, not `elif`, for clarity.

File: stdpopsim/cli.py [hunk 5]
```diff
@@ -708,20 +721,40 @@ def run_simulation(args):
...
+                if args.dfe_bed_file is not None:
+                    exit(
+                        "A DFE bed file and a DFE annotation have been "
+                        "selected. Please only use one."
+                    )
                 annot = species.get_annotations(args.dfe_annotation)
                 intervals = annot.get_chromosome_annotations(args.chromosome)
                 intervals_summary_str = f"{annot.id} elements on {args.chromosome}"
+            elif args.dfe_bed_file is not None:
+                intervals = np.loadtxt(args.dfe_bed_file, usecols=[1, 2], dtype="int")
```
Comment (petrelharp) [inline stdpopsim/cli.py:754]: did you check if the error it gives for a malformed BED file (eg accidentally calling it on something totally different) is interpretable?

File: tests/test_slim_engine.py [hunk 6]
```diff
@@ -421,6 +423,70 @@ def count_mut_types(ts):
             ).split()
             with pytest.raises(SystemExit):
                 capture_output(stdpopsim.cli.stdpopsim_main, cmd)
+        # test DFE with demography and interval as a bed file
```
Comment (petrelharp) [inline tests/test_slim_engine.py:426]: let's split this out into a separate `test_` function?

File: tests/test_slim_engine.py [hunk 7]
```diff
@@ -421,6 +423,70 @@ def count_mut_types(ts):
...
+                "-d OutOfAfrica_3G09 --dfe Gamma_K17 "
+                "--dfe-bed-file ex.bed 10"
+            ).split()
+            capture_output(stdpopsim.cli.stdpopsim_main, cmd)
+            ts = tskit.load(f.name)
+        assert ts.num_samples == 10
+        assert all(tree.num_roots == 1 for tree in ts.trees())
+        n_mut_types = count_mut_types(ts)
+        assert n_mut_types[0] > 0
+        assert n_mut_types[1] > 0
```
Comment (petrelharp) [inline tests/test_slim_engine.py:445]: We should check using top-level metadata reported from SLiM whether the intervals are doing what we think they should here. Want me ot take a crack at this?

File: tests/test_slim_engine.py [hunk 8]
```diff
@@ -421,6 +423,70 @@ def count_mut_types(ts):
...
+        with tempfile.NamedTemporaryFile(mode="w") as f:
+            cmd = (
+                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
+                f"HomSap -c chr22 -l0.01 -o {f.name} "
+                "-d OutOfAfrica_3G09 --dfe Gamma_K17 "
+                "--dfe-bed-file ex.bed "
+                "--dfe-annotation ensembl_havana_104_exons 10"
+            ).split()
+            with pytest.raises(SystemExit):
+                capture_output(stdpopsim.cli.stdpopsim_main, cmd)
```
Comment (petrelharp) [inline tests/test_slim_engine.py:478]: To make sure it's giving the *right* errror:
```suggestion
            with pytest.raises(SystemExit, match=".*bed file and a DFE annotation"):
                capture_output(stdpopsim.cli.stdpopsim_main, cmd)
```

File: tests/test_slim_engine.py [hunk 9]
```diff
@@ -421,6 +423,70 @@ def count_mut_types(ts):
...
+        # test DFE with interval file and annotation
+        with tempfile.NamedTemporaryFile(mode="w") as f:
+            cmd = (
+                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
+                f"HomSap -c chr22 -l0.01 -o {f.name} "
+                "-d OutOfAfrica_3G09 --dfe Gamma_K17 "
+                "--dfe-interval 999,1000 "
+                "--dfe-annotation ensembl_havana_104_exons 10"
+            ).split()
+            with pytest.raises(SystemExit):
```
Comment (petrelharp) [inline tests/test_slim_engine.py:466]: should check error is the right one; see below

**Comment by petrelharp** [comment]
Looks good, some fairly minor comments (but some rearranging of tests?).

File: tests/test_slim_engine.py (hunk 7 above)
Comment (izabelcavassim) [reply tests/test_slim_engine.py:445]: sure, go ahead :)

Comment (petrelharp) [reply tests/test_slim_engine.py:445]: ok, maybe let's get this PR merged and open an issue about that.

File: stdpopsim/cli.py (hunk 5 above)
Comment (izabelcavassim) [reply stdpopsim/cli.py:754]: No, doing it now

File: tests/test_slim_engine.py (hunk 9 above)
Comment (izabelcavassim) [reply tests/test_slim_engine.py:466]: done

File: tests/test_slim_engine.py (hunk 8 above)
Comment (izabelcavassim) [reply tests/test_slim_engine.py:478]: done

File: tests/test_slim_engine.py (hunk 6 above)
Comment (izabelcavassim) [reply tests/test_slim_engine.py:426]: I kind of think it is fine to have everything in one function, but lets just do this modification when we take this out of the test_slim_engine.py and transfer it to test_cli.py

File: stdpopsim/cli.py (hunk 4 above)
Comment (izabelcavassim) [reply stdpopsim/cli.py:724]: done

File: tests/test_slim_engine.py (hunk 6 above)
Comment (izabelcavassim) [reply tests/test_slim_engine.py:426]: Ah I see what you mean, maybe later? :S

File: stdpopsim/cli.py (hunk 5 above)
Comment (izabelcavassim) [reply stdpopsim/cli.py:754]: ok, it is something like:
OSError: "bla.bed" not found.
so interpretable, do we need to test this?

Comment (petrelharp) [reply stdpopsim/cli.py:754]: I was thinking a file that exists but is the wrong format - but we can open a new issue for this if we want.

File: tests/test_slim_engine.py (hunk 9 above)
Comment (petrelharp) [reply tests/test_slim_engine.py:470]: I see it's done below, but not here, I think?

**Comment by petrelharp** [comment]
Awesome, thanks! A quick squash/rebase and then we're good?

And, sorry to be nitpicky about checking the errors - it's just that weird errors can be a major source of user difficulty, and it's happened to me a bunch that the nice error message I wrote wasn't getting shown because some other error happened first.

**Comment by izabelcavassim** [comment]
Don't need to apologize @petrelharp! I like how detailed this is, and I am always learning a lot! I think you can merge now.

**Comment by izabelcavassim** [comment]
Can you merge this @petrelharp? Thanks

**Pull request merged**